File: 1401-1500/1405-Longest Happy String/1405-Longest Happy String.py
class Solution:
    def longestDiverseString(self, a: int, b: int, c: int) -> str:
        result = []
        currA = currB = currC = 0
        while a or b or c:
            if (a >= max(b, c) and currA < 2) or (a > 0 and (currB == 2 or currC == 2)):
                result.append('a')
                a -= 1
                currA += 1
                currB = currC = 0
            elif (b >= max(a, c) and currB < 2) or (b > 0 and (currA == 2 or currC == 2)):
                result.append('b')
                b -= 1
                currB += 1
                currA = currC = 0
            elif (c >= max(a, b) and currC < 2) or (c > 0 and (currA == 2 or currB == 2)):
                result.append('c')
                c -= 1
                currC += 1
                currA = currB = 0
            else:
                break
        return ''.join(result)

File: 1401-1500/1405-Longest Happy String/1405-Longest Happy String.cpp
class Solution {
public:
    string longestDiverseString(int a, int b, int c) {
        int currA = 0, currB = 0, currC = 0;
        string result;
        while (a || b || c) {
            if ((a >= max(b, c) && currA < 2) || (a > 0 && (currB == 2 || currC == 2))) {
                result += 'a';
                --a;
                ++currA;
                currB = currC = 0;
            }
            else if ((b >= max(a, c) && currB < 2) || (b > 0 && (currA == 2 || currC == 2))) {
                result += 'b';
                --b;
                ++currB;
                currA = currC = 0;
            }
            else if ((c >= max(a, b) && currC < 2) || (c > 0 && (currA == 2 || currB == 2))) {
                result += 'c';
                --c;
                ++currC;
                currA = currB = 0;
            }
            else {
                break;
            }
        }

        return result;
    }
};
